src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioRazorParser.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.VisualStudio.Text;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    public class VisualStudioRazorParser : IDisposable
+    {
+        private BackgroundParser _parser;
+        private RazorSyntaxTreePartialParser _partialParser;
+        private ITextBuffer _textBuffer;
+        private RazorCodeDocument _currentDocument;
+        private SourceChange _pendingShimTextChange;
+
+        public VisualStudioRazorParser(RazorTemplateEngine templateEngine, string filePath)
+        {
+            if (templateEngine == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngine));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(
+                    AspNetCore.Razor.Language.Resources.ArgumentCannotBeNullOrEmpty,
+                    nameof(filePath));
+            }
+
+            TemplateEngine = templateEngine;
+            FilePath = filePath;
+            _parser = new BackgroundParser(templateEngine, filePath);
+            _parser.ResultsReady += OnParseCompleted;
+
+            _parser.Start();
+        }
+
+        public event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        public event EventHandler ReparseWhenIdle;
+
+        public RazorTemplateEngine TemplateEngine { get; }
+
+        public string FilePath { get; }
```

Comment:
We need **something** that will give us the current state of the document. That's 60% of the reason why we're doing this.